core/unit_test/TestViewCopy.hpp

```diff
@@ -0,0 +1,134 @@
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <cstdio>
+
+#include <gtest/gtest.h>
+
+#include <Kokkos_Core.hpp>
+
+namespace Test {
+
+namespace {
+
+template < typename ExecSpace >
+struct TestViewCopy {
+
+  using InExecSpace = ExecSpace;
+
+  static void test_view_copy()
+  {
+#if defined( KOKKOS_ENABLE_CUDA ) || defined( KOKKOS_ENABLE_ROCM )
+   // ExecSpace = CudaUVM, CudaHostPinned
+   // This test will fail at runtime with an illegal memory access if something goes wrong
+   // Test 1: deep_copy from host_space to ExecSpace and ExecSpace back to host_space
+   {
+    const int dim0 = 4;
+    const int dim1 = 2;
+    const int dim2 = 3;
+
+    typedef Kokkos::View<double****,InExecSpace> Rank4ViewType;
+    Rank4ViewType view_4;
+    view_4 = Rank4ViewType("view_4", dim0, dim1, dim2, dim2);
+
+    typedef typename Kokkos::Impl::is_space<InExecSpace>::host_mirror_space::execution_space host_space_type;
+    Kokkos::View<double**,Kokkos::LayoutLeft,host_space_type> srcView("srcView", dim2, dim2);
+
+    // Strided dst view
+    auto dstView = Kokkos::subview(view_4, 0, 0, Kokkos::ALL(), Kokkos::ALL());
+
+    const Kokkos::pair<int,int> range(0, dim2);
+
+    // host_space to ExecSpace
+    Kokkos::deep_copy( dstView, srcView );
```

Comment:
@crtrott this should test the host -> X space

Reply:
@crtrott oh, I see what you're saying. I used the host mirror to get the host space, but that won't be sufficient to test for example Kokkos::Serial as a host space. I'll add tests for this.

Reply:
This is still not enough. The problem is that for UVM the HostMirror is also UVM. So you are only testing UVM->UVM. But the failure happened for CudaUVMSpace->HostSpace or vice versa.

Reply:
@crtrott just pushed updates to the unit tests so that either Serial, OpenMP, or Threads is used as a host space for testing (not just the host mirror space). Posted passing spot-check results below.